### ISSUE TITLE
Autosave can result in lost of data

### ISSUE THREAD

**Issue by adessein** [retext-project/retext#461]
Hi,

I have just experienced a loss of data on two files. FILE1 was an old file opened but I was not working on it. FILE2 was a new file that I was modifing when my computer crashed (not due to Retext).

When I restarted my computer both files were empty. What is frustrating is that FILE1 has been saved at some point, it is not like I have never saved. FILE2 has actually just been opened, not modified at all, still, his content vanished.

I had autosave enabled.

I think what happened is that retext has automatically saved the files with an empty content (for some reason...).

I think you should reconsider the autosave feature. It is dangerous to save on the original file without the user content. And I can see that some people are confused by the fact that the save button is disabled. It should behave like an auto-backup, a feature often seen in other programs. It should save the changes into a temporary file until the user actually saves the file. Then the temporary file is deleted. If the computer crashes and the program restarts, the existence of the temporary file is detected and the user is given the choice to open the backup instead.

**Comment by mitya57** [comment]
I am sorry that you had data loss because of ReText.

ReText writes files in order, not simultaneously. So when your computer crashed, at least one of the files should have been saved on disk. Maybe this is a filesystem issue? If it had not been unmounted properly, then data loss could occur. And in this case backup copies would not help…

I have just pushed a commit that makes files auto-saved only when they are modified. This should avoid re-saving of opened but not modified files.

I am not opposed to the idea of backup copies, I just don't see how they would have helped in your particular case.

**Comment by adessein** [comment]
Hi mitya57,

Thank you for you quick reaction to my issue. What happened is a bit weird and I am not sure I explained well. Just to make sure you understand, I will rephrase.

I had two instances of ReText opened, two different files. I was only editing one of them. After I restarted my computer after it froze, both files were empty. As if ReText had autosaved with just nothing. And it did on both files, even the one that was not modified.

Therefore I think your commit could help for the non-modified file. However, the root cause is still hard to understamd: why did ReText saved an empty file over the existing ones?

Maybe you are right, it is not ReText, it is the filesystem. It is hard to say.

**Comment by mitya57** [comment]
To know the root cause, I need some way to reproduce it.

I tried killing ReText process when auto-save is enabled, in all cases the file was saved correctly (not having last-minute changes, but not empty).

**Comment by JanKanis** [comment]
Hi @adessein, this sounds like it is filesystem related. It would help if you could post what operating system and version you use, and what filesystem and (if non-standard) what mount options. 

However ReText also does not appear to do the safest thing. Filesystems often don't write their data to disk immediately. To force that, ReText should call `fsync` on the file after writing to it, that will force the filesystem to write all data immediately. (`savestream.flush(); os.fsync(savefile.handle())`) That still leaves a small interval between clearing the old content and writing the new content. What many editors do is save the file content in a temporary file (and `fsync` that), which can be deleted after the original file has been successfully written to and fsynced. On startup, the editor can check if any such temporary files exist, if that is the case it offers the user to open it so the user can save again if wanted.

**Comment by internationils** [comment]
I would agree that having autosave saving backups would be valuable. Something like filename.backup.pid ... using the process ID in case multiple instances of retext are open (possibly even on multiple hosts with dropbox or similar shared folders). I would rather remove left over backups and clean up after a crash than risk data loss.
This would go hand in hand with https://github.com/retext-project/retext/issues/517 to allow editing on multiple systems.

**Comment by djowett** [comment]
I just found half of a file gone - my guess is this happened like this:

1. inadvertantly select half the file and press a key
2. autosave triggers
3. close the app without realizing the above

Probably I should turn off auto-save to avoid this?

But I'd like to see some kind of rolling backup algorithm that saves the last X versions of a file (maybe on application startup?)

**Comment by djowett** [comment]
Strangely I found the file recovered today, which is a great relief! I'm not sure why though, perhaps it was only half displayed yesterday for some reason :shrug: 

Rolling backup would still be nice to not feel 1 step away from disaster
